webrender/src/device/gfx/device.rs

```diff
@@ -1341,41 +1351,95 @@ impl<B: hal::Backend> Device<B> {
                 self.fbos[&self.bound_draw_fbo].format,
+                if rbo_id == RBOId(0) {
+                    (None, false)
+                } else {
+                    let mut depth_test_changed =  false;
+                    // This is needed to avoid validation layer errors for different attachments between frambuffer renderpass and pipelinelayout
+                    if self.current_depth_test == DepthTest::Off {
+                        self.current_depth_test = LESS_EQUAL_TEST;
+                        depth_test_changed = true;
+                    }
+                    (Some(&self.rbos[&rbo_id].core), depth_test_changed)
+                },
             )
         } else {
             if self.current_depth_test == DepthTest::Off {
                 (
+                    &self.frame_images[self.current_frame_id],
                     &self.framebuffers[self.current_frame_id],
                     self.surface_format,
+                    (None, false)
                 )
             } else {
                 (
+                    &self.frame_images[self.current_frame_id],
                     &self.framebuffers_depth[self.current_frame_id],
                     self.surface_format,
+                    (Some(&self.frame_depths[self.current_frame_id].core), false)
                 )
             }
         };
         let rp = self
             .render_pass
             .as_ref()
             .unwrap()
-            .get_render_pass(format, self.current_depth_test != DepthTest::Off);
+            .get_render_pass(format, depth_img.is_some());
+
+        let before_state = img.state.get();
+        let mut before_depth_state = None;
+        let mut pre_stage = Some(PipelineStage::empty());
+        let cmd_buffer = self.command_pool[self.next_id].acquire_command_buffer();
+        unsafe {
+            cmd_buffer.begin();
+            if let Some(barrier) = img.transit(
+                hal::image::Access::empty(),
+                hal::image::Layout::ColorAttachmentOptimal,
+                img.subresource_range.clone(),
+                pre_stage.as_mut(),
+            ) {
+                cmd_buffer.pipeline_barrier(
+                    pre_stage.unwrap().. PipelineStage::COLOR_ATTACHMENT_OUTPUT,
+                    hal::memory::Dependencies::empty(),
+                    &[barrier],
+                );
+            }
+            if let Some(depth_img) = depth_img {
+                before_depth_state = Some(depth_img.state.get());
+                if let Some(barrier) = depth_img.transit(
+                    hal::image::Access::empty(),
+                    hal::image::Layout::DepthStencilAttachmentOptimal,
+                    depth_img.subresource_range.clone(),
+                   None,
+                ) {
+                    cmd_buffer.pipeline_barrier(
+                        PipelineStage::EARLY_FRAGMENT_TESTS
+                            .. PipelineStage::LATE_FRAGMENT_TESTS,
```

Comment:
I think these barriers are incorrect. When you are saying that this depth image needs to get DepthStencilAttachmentOptimal layout, you want this to happen before this image starts to be used as a d/s attachment. So your barrier should end at EARLY_FRAGMENT_TESTS (and start point would depend on what attachment/use it was in, previously)